solidity/contracts/AugurConstantProductMarket.sol

```diff
@@ -12,186 +12,107 @@ import { AddressToString } from "./AddressToString.sol";
 contract AugurConstantProduct is ERC20 {
 	using AddressToString for address;
 
-
-	IERC20 public dai = ERC20(Constants.DAI_ADDRESS);
 	IShareToken public shareToken;
 	address public augurMarketAddress;
 	IAugur public constant augur = IAugur(Constants.AUGUR_ADDRESS);
-	uint256 public numTicks;
 	uint256 public INVALID;
 	uint256 public NO;
 	uint256 public YES;
-	uint256 public constant afterFeeDenominator = 1000;
-	uint256 public constant afterFeeNumerator = 950; // 5% fee
+	uint256 public noBalance;
+	uint256 public yesBalance;
+	uint256 public constant feeScope = 1000;
+	uint256 public constant fee = 50; // 5% fee
+
+	uint private unlocked = 1;
+	modifier lock() {
+		require(unlocked == 1, 'LOCKED');
+		unlocked = 0;
+		_;
+		unlocked = 1;
+	}
 
 	constructor(IMarket market) ERC20(string(abi.encodePacked("ACPM-", address(market).addressToString())), address(market).addressToString()) {
 		augurMarketAddress = address(market);
 		shareToken = market.shareToken();
-		numTicks = market.getNumTicks();
-		dai.approve(address(shareToken), 2**256-1);
-		dai.approve(Constants.AUGUR_ADDRESS, 2**256-1);
-		require(augur.getMarketType(market) == 0, "ACPM only supports Yes No Markets");
+		require(augur.getMarketType(market) == 0, "AugurCP: ACPM only supports Yes No Markets");
 		INVALID = shareToken.getTokenId(augurMarketAddress, 0);
 		NO = shareToken.getTokenId(augurMarketAddress, 1);
 		YES = shareToken.getTokenId(augurMarketAddress, 2);
 	}
 
-	function addLiquidity(uint256 sharesToBuy) external {
-		// TODO: gas golf this function a bit, though it doesn't really matter
-		uint256 poolConstantBefore = sqrt(poolConstant());
+	function mint(address mintTo) lock external {
+		(uint256 poolNo, uint256 poolYes) = noYesShareBalances(address(this));
+		uint256 noIn = poolNo - noBalance;
+		uint256 yesIn = poolYes - yesBalance;
 
-		dai.transferFrom(msg.sender, address(this), sharesToBuy * numTicks);
-		shareToken.buyCompleteSets(augurMarketAddress, address(this), sharesToBuy);
+		uint256 supply = totalSupply();
+		uint256 liquidity = 0;
 
-		if (poolConstantBefore == 0) {
-			_mint(msg.sender, sqrt(poolConstant()));
+		if (supply == 0) {
+			liquidity = sqrt(noIn * yesIn); // CONSIDER: MINIMUM BALANCE handling?
 		} else {
-			_mint(msg.sender, totalSupply() * sqrt(poolConstant()) / poolConstantBefore - totalSupply());
+			uint256 liquidity1 = noIn * supply / noBalance;
+			uint256 liquidity2 = yesIn * supply / yesBalance;
+			liquidity = liquidity1 < liquidity2 ? liquidity1 : liquidity2;
 		}
+
+		_mint(mintTo, liquidity);
+
+		noBalance = poolNo;
+		yesBalance = poolYes;
```

Comment:
there's a risk of bug to manage these share balances internally, as you can send extra balance to the contract and then the pool has more of assets than its accounting for

Reply:
Won't really manifest as a bug, just someone losing their money. Same as how you can send a token to a univ2 pair.